rust/sql/src/postgres/reader.rs

```diff
@@ -0,0 +1,894 @@
+                return Err(ArrowError::IoError(
+                    "table name must have schema and table name only, or just table name"
+                        .to_string(),
+                ));
+            }
+            (
+                format!("table_schema = '{}'", split[0]),
+                format!(" and table_name = '{}'", split[1]),
+            )
+        } else {
+            (String::new(), format!("table_name = '{}'", table_name))
+        };
+        let mut client = Client::connect(connection_string, NoTls).unwrap();
+        let results = client
+            .query(format!("select column_name, ordinal_position, is_nullable, data_type, character_maximum_length, numeric_precision, datetime_precision from information_schema.columns where {}{}", table_schema, table_name).as_str(), &[])
+            .unwrap();
+        let fields: Result<Vec<Field>> = results
+            .iter()
+            .map(|row| PgDataType {
+                column_name: row.get("column_name"),
+                ordinal_position: row.get("ordinal_position"),
+                is_nullable: row.get("is_nullable"),
+                data_type: row.get("data_type"),
+                char_max_length: row.get("character_maximum_length"),
+                numeric_precision: row.get("numeric_precision"),
+                datetime_precision: row.get("datetime_precision"),
+            })
+            .map(Field::try_from)
+            .collect();
+        Ok(Schema::new(fields.unwrap()))
+    }
+
+    fn read_table(
+        connection: &str,
+        table_name: &str,
+        limit: Option<usize>,
+        _batch_size: usize,
+    ) -> Result<Vec<RecordBatch>> {
+        // create connection
+        let mut client = Client::connect(connection, NoTls).unwrap();
+        let total_rows = client
+            .query_one(
+                format!("select count(1) as freq from {}", table_name).as_str(),
+                &[],
+            )
+            .expect("Unable to get row count");
+        let total_rows: i64 = total_rows.get("freq");
+        let limit = limit.unwrap_or(std::usize::MAX).min(total_rows as usize);
+        // get schema
+        // TODO: reuse connection
+        // TODO: split read into multiple batches, using limit and skip
+        let schema = Postgres::get_table_schema(connection, table_name)?;
+        let reader = get_binary_reader(
+            &mut client,
+            format!("select * from {} limit {}", table_name, limit).as_str(),
+        )?;
+        read_from_binary(reader, &schema).map(|batch| vec![batch])
+    }
+
+    fn read_query(
+        connection: &str,
+        query: &str,
+        limit: Option<usize>,
+        _batch_size: usize,
+    ) -> Result<Vec<RecordBatch>> {
+        // create connection
+        let mut client = Client::connect(connection, NoTls).unwrap();
+        let total_rows = client
+            .query_one(
+                format!("select count(1) as freq from ({}) a", query).as_str(),
```

Comment:
We'd likely need some way of determining whether this is cheap enough to call. The idea's that we determine the number of records, so that we can split the query into subqueries with `limit` and `skip`.
A query engine could do this globally, and then partition the overall query into `n` readers.